File: src/apiconvert-core/src/rules-bundler.ts
import fs from "node:fs";
import path from "node:path";
import { normalizeConversionRules } from "./rules-normalizer";
import type { ConversionRules, RuleBundleOptions } from "./types";

type RawRulesFile = {
  include?: unknown;
  inputFormat?: unknown;
  outputFormat?: unknown;
  rules?: unknown;
};

export function bundleConversionRules(entryRulesPath: string, options: RuleBundleOptions = {}): ConversionRules {
  const baseDirectory = path.resolve(options.baseDirectory ?? path.dirname(entryRulesPath));
  const entryPath = path.resolve(baseDirectory, entryRulesPath);
  const visited = new Set<string>();

  return bundleFile(entryPath, visited, []);
}

function bundleFile(filePath: string, visited: Set<string>, chain: string[]): ConversionRules {
  const normalizedFilePath = path.resolve(filePath);
  if (chain.includes(normalizedFilePath)) {
    const circular = [...chain, normalizedFilePath].map((entry) => path.basename(entry)).join(" -> ");
    throw new Error(`Circular include detected: ${circular}`);
  }

  if (visited.has(normalizedFilePath)) {
    return normalizeConversionRules({ inputFormat: "json", outputFormat: "json", rules: [] });
  }

  visited.add(normalizedFilePath);
  const raw = loadRawRulesFile(normalizedFilePath);

  const includes = normalizeIncludes(raw.include);
  const currentChain = [...chain, normalizedFilePath];
  const includedRules: ConversionRules[] = includes.map((includePath) => {
    const resolvedIncludePath = path.resolve(path.dirname(normalizedFilePath), includePath);
    return bundleFile(resolvedIncludePath, visited, currentChain);
  });

  const localRules = normalizeConversionRules({
    inputFormat: raw.inputFormat,
    outputFormat: raw.outputFormat,
    rules: raw.rules
  });

  const mergedRules = includedRules.flatMap((entry) => entry.rules ?? []).concat(localRules.rules ?? []);
  const inheritedFormat = includedRules.find((entry) => entry.inputFormat != null)?.inputFormat;
  const inheritedOutputFormat = includedRules.find((entry) => entry.outputFormat != null)?.outputFormat;

  return {
    inputFormat: localRules.inputFormat ?? inheritedFormat,
    outputFormat: localRules.outputFormat ?? inheritedOutputFormat,
    rules: mergedRules,
    validationErrors: [
      ...(includedRules.flatMap((entry) => entry.validationErrors ?? [])),
      ...(localRules.validationErrors ?? [])
    ]
  };
}

function loadRawRulesFile(filePath: string): RawRulesFile {
  if (!fs.existsSync(filePath)) {
    throw new Error(`Included rules file not found: ${filePath}`);
  }

  const text = fs.readFileSync(filePath, "utf8");
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (error) {
    throw new Error(
      `Invalid JSON in rules file ${filePath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }

  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error(`Rules file must contain a JSON object: ${filePath}`);
  }

  return parsed as RawRulesFile;
}

function normalizeIncludes(includeValue: unknown): string[] {
  if (includeValue == null) {
    return [];
  }

  if (!Array.isArray(includeValue)) {
    throw new Error("include must be an array of relative file paths.");
  }

  return includeValue
    .filter((entry): entry is string => typeof entry === "string")
    .map((entry) => entry.trim())
    .filter((entry) => entry.length > 0)
    .sort((a, b) => a.localeCompare(b));
}

File: src/Apiconvert.Core/Converters/RulesBundler.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Apiconvert.Core.Rules;

namespace Apiconvert.Core.Converters;

internal static class RulesBundler
{
    internal static ConversionRules BundleRules(string entryRulesPath, RuleBundleOptions? options = null)
    {
        if (string.IsNullOrWhiteSpace(entryRulesPath))
        {
            throw new ArgumentException("Entry rules path is required.", nameof(entryRulesPath));
        }

        var baseDirectory = string.IsNullOrWhiteSpace(options?.BaseDirectory)
            ? Directory.GetCurrentDirectory()
            : Path.GetFullPath(options.BaseDirectory);
        var entryPath = Path.GetFullPath(Path.Combine(baseDirectory, entryRulesPath));
        var visited = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        return BundleFile(entryPath, visited, []);
    }

    private static ConversionRules BundleFile(
        string filePath,
        HashSet<string> visited,
        IReadOnlyList<string> chain)
    {
        var normalizedFilePath = Path.GetFullPath(filePath);
        if (chain.Contains(normalizedFilePath, StringComparer.OrdinalIgnoreCase))
        {
            var circular = string.Join(" -> ", chain.Append(normalizedFilePath).Select(Path.GetFileName));
            throw new InvalidOperationException($"Circular include detected: {circular}");
        }

        if (visited.Contains(normalizedFilePath))
        {
            return new ConversionRules();
        }

        visited.Add(normalizedFilePath);

        var raw = LoadRawRulesFile(normalizedFilePath);
        var includes = NormalizeIncludes(raw["include"]);
        var nextChain = chain.Append(normalizedFilePath).ToArray();

        var includedRules = new List<ConversionRules>();
        foreach (var include in includes)
        {
            var includeBase = Path.GetDirectoryName(normalizedFilePath) ?? string.Empty;
            var includePath = Path.GetFullPath(Path.Combine(includeBase, include));
            includedRules.Add(BundleFile(includePath, visited, nextChain));
        }

        var localObject = new JsonObject
        {
            ["inputFormat"] = raw["inputFormat"]?.DeepClone(),
            ["outputFormat"] = raw["outputFormat"]?.DeepClone(),
            ["rules"] = raw["rules"]?.DeepClone()
        };
        var localRules = RulesNormalizer.NormalizeConversionRules(localObject.ToJsonString());

        return new ConversionRules
        {
            InputFormat = localRules.InputFormat,
            OutputFormat = localRules.OutputFormat,
            Rules = includedRules.SelectMany(entry => entry.Rules).Concat(localRules.Rules).ToList(),
            ValidationErrors = includedRules
                .SelectMany(entry => entry.ValidationErrors)
                .Concat(localRules.ValidationErrors)
                .ToList()
        };
    }

    private static JsonObject LoadRawRulesFile(string filePath)
    {
        if (!File.Exists(filePath))
        {
            throw new FileNotFoundException($"Included rules file not found: {filePath}");
        }

        var text = File.ReadAllText(filePath);
        JsonNode? node;
        try
        {
            node = JsonNode.Parse(text);
        }
        catch (JsonException ex)
        {
            throw new InvalidOperationException($"Invalid JSON in rules file {filePath}: {ex.Message}", ex);
        }

        if (node is not JsonObject jsonObject)
        {
            throw new InvalidOperationException($"Rules file must contain a JSON object: {filePath}");
        }

        return jsonObject;
    }

    private static List<string> NormalizeIncludes(JsonNode? includeNode)
    {
        if (includeNode is null)
        {
            return [];
        }

        if (includeNode is not JsonArray includeArray)
        {
            throw new InvalidOperationException("include must be an array of relative file paths.");
        }

        return includeArray
            .Select(node => node?.GetValue<string>()?.Trim())
            .Where(value => !string.IsNullOrWhiteSpace(value))
            .Cast<string>()
            .Distinct(StringComparer.Ordinal)
            .OrderBy(value => value, StringComparer.Ordinal)
            .ToList();
    }
}
